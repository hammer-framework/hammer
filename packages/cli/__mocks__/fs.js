import { sep } from 'path'

import { isString } from 'lodash'

const fs = {
  ...jest.requireActual('fs'),
}

// This is a custom function that our tests can use during setup to specify
// what the files on the "mock" filesystem should look like when any of the
// `fs` APIs are used.
let mockFiles = {}
/** @param newMockFiles - {[filepath]: contents} */
fs.__setMockFiles = (newMockFiles) => {
  mockFiles = { ...newMockFiles }
}

fs.existsSync = (path) => {
  return isString(mockFiles[path])
}

fs.mkdirSync = () => {}

fs.readFileSync = (path) => {
  // In prisma v4.3.0, prisma format uses a Wasm module. See https://github.com/prisma/prisma/releases/tag/4.3.0.
  // We shouldn't mock this, so we'll use the real fs.readFileSync.
  if (path.includes('prisma_fmt_build_bg.wasm')) {
    return jest.requireActual('fs').readFileSync(path)
  }

  return mockFiles[path]
}

fs.readdirSync = (path) => {
  const mockedFiles = []
  Object.keys(mockFiles).forEach((mockedPath) => {
    // Is a child of the desired path
    if (mockedPath.startsWith(path)) {
      const childPath = mockedPath.substring(path.length + 1)
      // Is the child a direct child of the desired path
      if (!childPath.includes(sep)) {
        mockedFiles.push(childPath)
      }
    }
  })
  return mockedFiles
}

fs.writeFileSync = (path, contents) => {
  mockFiles[path] = contents
}

fs.unlinkSync = (path) => {
  delete mockFiles[path]
}

fs.rmSync = (path) => {
  delete mockFiles[path]
}

// @MARK this works for now but should probably be a little more full-featured
fs.rmdirSync = (path) => {
  delete mockFiles[path]
}

module.exports = fs
