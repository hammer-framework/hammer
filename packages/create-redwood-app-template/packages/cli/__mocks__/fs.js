export const readFileSync = () => {
  console.log('hello')

  return ''
}
