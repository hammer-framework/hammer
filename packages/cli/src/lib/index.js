import fs from 'fs'
import path from 'path'

import lodash from 'lodash/string'
import camelcase from 'camelcase'
import pascalcase from 'pascalcase'
import pluralize from 'pluralize'
import { paramCase } from 'param-case'
import { getDMMF } from '@prisma/sdk'
import { getPaths as getRedwoodPaths } from '@redwoodjs/internal'
import execa from 'execa'
import Listr from 'listr'
import VerboseRenderer from 'listr-verbose-renderer'

import c from 'src/lib/colors'

export const asyncForEach = async (array, callback) => {
  for (let index = 0; index < array.length; index++) {
    await callback(array[index], index, array)
  }
}

/**
 * Returns the database schema for the given `name` database table parsed from
 * the schema.prisma of the target applicaiton. If no `name` is given then the
 * entire schema is returned.
 */
export const getSchema = async (name) => {
  const schemaPath = path.join(getPaths().api.db, 'schema.prisma')
  const metadata = await getDMMF({
    datamodel: readFile(schemaPath.toString()),
  })

  if (name) {
    const model = metadata.datamodel.models.find((model) => {
      return model.name === name
    })

    if (model) {
      return model
    } else {
      throw `No schema definition found for \`${name}\``
    }
  }

  return metadata.datamodel
}

/**
 * Returns variants of the passed `name` for usage in templates. If the given
 * name was "fooBar" then these would be:

 * pascalName: FooBar
 * singularPascalName: FooBar
 * pluralPascalName: FooBars
 * singularCamelName: fooBar
 * pluralCamelName: fooBars
 * singularParamName: foo-bar
 * pluralParamName: foo-bars
*/
const nameVariants = (name) => {
  const normalizedName = pascalcase(pluralize.singular(name))

  return {
    pascalName: pascalcase(name),
    camelName: camelcase(name),
    singularPascalName: normalizedName,
    pluralPascalName: pluralize(normalizedName),
    singularCamelName: camelcase(normalizedName),
    pluralCamelName: camelcase(pluralize(normalizedName)),
    singularParamName: paramCase(normalizedName),
    pluralParamName: paramCase(pluralize(normalizedName)),
  }
}

export const templateRoot = path.resolve(
  __dirname,
  '../commands/Generate/templates'
)

export const generateTemplate = (templateFilename, { name, ...rest }) => {
  const templatePath = path.join(templateRoot, templateFilename)
  const template = lodash.template(readFile(templatePath).toString())
  return template({
    name,
    ...nameVariants(name),
    ...rest,
  })
}

export const readFile = (target) => fs.readFileSync(target)

export const writeFile = async (
  target,
  contents,
  { overwriteExisting = false } = {}
) => {
  if (!overwriteExisting && fs.existsSync(target)) {
    throw new Error(`${target} already exists.`)
  }

  const filename = path.basename(target)
  const targetDir = target.replace(filename, '')
  fs.mkdirSync(targetDir, { recursive: true })
  fs.writeFileSync(target, contents)
}

export const bytes = (contents) => Buffer.byteLength(contents, 'utf8')

/**
 * This wraps the core version of getPaths into something that catches the exception
 * and displays a helpful error message.
 */
export const getPaths = () => {
  try {
    return getRedwoodPaths()
  } catch (e) {
    console.log(c.error(e.message))
    process.exit(0)
  }
}

/**
 * Creates a list of tasks that write files to the disk.
 *
 * @param files - {[filepath]: contents}
 */
export const writeFilesTask = (files, options) => {
  const { base } = getPaths()
  return new Listr(
    Object.keys(files).map((file) => {
      const contents = files[file]
      return {
        title: `Writing \`./${path.relative(base, file)}\`...`,
        task: () => writeFile(file, contents, options),
      }
    })
  )
}

/**
 * Update the project's routes file.
 */
export const addRoutesToRouterTask = (routes) => {
  const redwoodPaths = getPaths()
  const routesContent = readFile(redwoodPaths.web.routes).toString()
  const newRoutesContent = routes.reduce((content, route) => {
    return content.replace(/(\s*)\<Router\>/, `$1<Router>$1  ${route}`)
  }, routesContent)
  writeFile(redwoodPaths.web.routes, newRoutesContent, {
    overwriteExisting: true,
  })
}

export const runCommandTask = async (commands, { verbose }) => {
  const tasks = new Listr(
    commands.map(({ title, cmd, args, opts = {} }) => ({
      title,
      task: async () => {
        return execa(cmd, args, {
          shell: true,
          cwd: `${getPaths().base}/api`,
          stdio: verbose ? 'inherit' : 'pipe',
          extendEnv: true,
          cleanup: true,
          ...opts,
        })
      },
    })),
    {
      renderer: verbose && VerboseRenderer,
      dateFormat: false,
    }
  )

  try {
    await tasks.run()
  } catch (e) {
    console.log(c.error(e.message))
  }
}

export * from './generateTempSchema'
