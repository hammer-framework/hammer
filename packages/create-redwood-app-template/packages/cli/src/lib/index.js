import fs from 'fs'
import path from 'path'

import lodash from 'lodash/string'
import camelcase from 'camelcase'
import pascalcase from 'pascalcase'
import pluralize from 'pluralize'
import { paramCase } from 'param-case'
import { getDMMF } from '@prisma/sdk'
import { getPaths as getRedwoodPaths } from '@redwoodjs/internal'
import execa from 'execa'
import Listr from 'listr'
import VerboseRenderer from 'listr-verbose-renderer'
import { format } from 'prettier'

import c from 'src/lib/colors'

export const asyncForEach = async (array, callback) => {
  for (let index = 0; index < array.length; index++) {
    await callback(array[index], index, array)
  }
}

/**
 * Returns the database schema for the given `name` database table parsed from
 * the schema.prisma of the target applicaiton. If no `name` is given then the
 * entire schema is returned.
 */
export const getSchema = async (name) => {
  const schemaPath = path.join(getPaths().api.db, 'schema.prisma')
  const metadata = await getDMMF({
    datamodel: readFile(schemaPath.toString()),
  })

  if (name) {
    const model = metadata.datamodel.models.find((model) => {
      return model.name === name
    })

    if (model) {
      return model
    } else {
      throw `No schema definition found for \`${name}\` in schema.prisma file`
    }
  }

  return metadata.datamodel
}

/**
 * Returns variants of the passed `name` for usage in templates. If the given
 * name was "fooBar" then these would be:

 * pascalName: FooBar
 * singularPascalName: FooBar
 * pluralPascalName: FooBars
 * singularCamelName: fooBar
 * pluralCamelName: fooBars
 * singularParamName: foo-bar
 * pluralParamName: foo-bars
*/
const nameVariants = (name) => {
  const normalizedName = pascalcase(pluralize.singular(name))

  return {
    pascalName: pascalcase(name),
    camelName: camelcase(name),
    singularPascalName: normalizedName,
    pluralPascalName: pluralize(normalizedName),
    singularCamelName: camelcase(normalizedName),
    pluralCamelName: camelcase(pluralize(normalizedName)),
    singularParamName: paramCase(normalizedName),
    pluralParamName: paramCase(pluralize(normalizedName)),
  }
}

export const templateRoot = path.resolve(
  __dirname,
  '../commands/generate/templates'
)

export const generateTemplate = (templateFilename, { name, ...rest }) => {
  const templatePath = path.join(templateRoot, templateFilename)
  const template = lodash.template(readFile(templatePath).toString())

  const renderedTemplate = template({
    name,
    ...nameVariants(name),
    ...rest,
  })

  // We format .js and .css templates, we need to tell prettier which parser
  // we're using.
  // https://prettier.io/docs/en/options.html#parser
  const parser = {
    css: 'css',
    js: 'babel',
  }[path.extname(templateFilename)]

  if (typeof parser === 'undefined') {
    return renderedTemplate
  }

  return format(renderedTemplate, {
    ...prettierOptions(),
    parser,
  })
}

export const readFile = (target) => fs.readFileSync(target)

export const writeFile = async (
  target,
  contents,
  { overwriteExisting = false } = {}
) => {
  if (!overwriteExisting && fs.existsSync(target)) {
    throw new Error(`${target} already exists.`)
  }

  const filename = path.basename(target)
  const targetDir = target.replace(filename, '')
  fs.mkdirSync(targetDir, { recursive: true })
  fs.writeFileSync(target, contents)
}

export const bytes = (contents) => Buffer.byteLength(contents, 'utf8')

/**
 * This wraps the core version of getPaths into something that catches the exception
 * and displays a helpful error message.
 */
export const getPaths = () => {
  try {
    return getRedwoodPaths()
  } catch (e) {
    console.log(c.error(e.message))
    process.exit(0)
  }
}

/**
 * This returns the config present in `prettier.config.js` of a Redwood project.
 */
export const prettierOptions = () => {
  try {
    return require(path.join(getPaths().base, 'prettier.config.js'))
  } catch (e) {
    return undefined
  }
}

/**
 * Creates a list of tasks that write files to the disk.
 *
 * @param files - {[filepath]: contents}
 */
export const writeFilesTask = (files, options) => {
  const { base } = getPaths()
  return new Listr(
    Object.keys(files).map((file) => {
      const contents = files[file]
      return {
        title: `Writing \`./${path.relative(base, file)}\`...`,
        task: () => writeFile(file, contents, options),
      }
    })
  )
}

/**
 * Update the project's routes file.
 */
export const addRoutesToRouterTask = (routes) => {
  const redwoodPaths = getPaths()
  const routesContent = readFile(redwoodPaths.web.routes).toString()
  const newRoutesContent = routes.reverse().reduce((content, route) => {
    if (content.includes(route)) {
      return content
    }
    return content.replace(/(\s*)\<Router\>/, `$1<Router>$1  ${route}`)
  }, routesContent)
  writeFile(redwoodPaths.web.routes, newRoutesContent, {
    overwriteExisting: true,
  })
}

export const runCommandTask = async (commands, { verbose }) => {
  const tasks = new Listr(
    commands.map(({ title, cmd, args, opts = {} }) => ({
      title,
      task: async () => {
        return execa(cmd, args, {
          shell: true,
          cwd: `${getPaths().base}/api`,
          stdio: verbose ? 'inherit' : 'pipe',
          extendEnv: true,
          cleanup: true,
          ...opts,
        })
      },
    })),
    {
      renderer: verbose && VerboseRenderer,
      dateFormat: false,
    }
  )

  try {
    await tasks.run()
    return true
  } catch (e) {
    console.log(c.error(e.message))
    return false
  }
}
