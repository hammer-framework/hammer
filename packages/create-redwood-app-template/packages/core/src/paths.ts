import path from 'path'
import fs from 'fs'

import findUp from 'findup-sync'

import { Paths, PagesDependency } from './types'

const CONFIG_FILE_NAME = 'redwood.toml'

const PATH_API_DIR_FUNCTIONS = 'api/src/functions'
const PATH_API_DIR_GRAPHQL = 'api/src/graphql'
const PATH_API_DIR_DB = 'api/prisma'
const PATH_API_DIR_SERVICES = 'api/src/services'
const PATH_API_DIR_SRC = 'api/src'
const PATH_WEB_ROUTES = 'web/src/Routes.js'
const PATH_WEB_DIR_LAYOUTS = 'web/src/layouts/'
const PATH_WEB_DIR_PAGES = 'web/src/pages/'
const PATH_WEB_DIR_COMPONENTS = 'web/src/components'
const PATH_WEB_DIR_SRC = 'web/src'

/**
 * Search the parent directories for the Redwood configuration file.
 */
export const getConfigPath = (): string => {
  const configPath = findUp(CONFIG_FILE_NAME)
  if (!configPath) {
    throw new Error(
      `Could not find a "${CONFIG_FILE_NAME}" file, are you sure you're in a Redwood project?`
    )
  }
  return configPath
}

/**
 * The Redwood config file is used as an anchor for the base directory of a project.
 */
export const getBaseDir = (configPath: string = getConfigPath()): string => {
  return path.dirname(configPath)
}

/**
 * Path constants that are relevant to a Redwood project.
 */
export const getPaths = (BASE_DIR: string = getBaseDir()): Paths => {
  return {
    base: BASE_DIR,
    api: {
      db: path.join(BASE_DIR, PATH_API_DIR_DB),
      functions: path.join(BASE_DIR, PATH_API_DIR_FUNCTIONS),
      graphql: path.join(BASE_DIR, PATH_API_DIR_GRAPHQL),
      services: path.join(BASE_DIR, PATH_API_DIR_SERVICES),
      src: path.join(BASE_DIR, PATH_API_DIR_SRC),
    },
    web: {
      routes: path.join(BASE_DIR, PATH_WEB_ROUTES),
      pages: path.join(BASE_DIR, PATH_WEB_DIR_PAGES),
      components: path.join(BASE_DIR, PATH_WEB_DIR_COMPONENTS),
      layouts: path.join(BASE_DIR, PATH_WEB_DIR_LAYOUTS),
      src: path.join(BASE_DIR, PATH_WEB_DIR_SRC),
    },
  }
}

/**
 * Recursively process the pages directory and return information useful for
 * automated imports.
 */
export const processPagesDir = (
  webPagesDir: string = getPaths().web.pages,
  prefix: Array<string> = []
): Array<PagesDependency> => {
  const deps: Array<PagesDependency> = []
  const entries = fs.readdirSync(webPagesDir, { withFileTypes: true })

  // Iterate over a dir's entries, recursing as necessary into
  // subdirectories.
  entries.forEach((entry) => {
    if (entry.isDirectory()) {
      // Actual JS files reside in a directory of the same name, so let's
      // construct the filename of the actual Page file.
      const testFile = path.join(webPagesDir, entry.name, entry.name + '.js')

      if (fs.existsSync(testFile)) {
        // If the Page exists, then construct the dependency object and push it
        // onto the deps array.
        const basename = path.posix.basename(entry.name, '.js')
        const importName = prefix.join() + basename
        const importFile = path.join('src', 'pages', ...prefix, basename)
        deps.push({
          const: importName,
          path: path.join(webPagesDir, entry.name),
          importStatement: `const ${importName} = { name: '${importName}', loader: () => import('${importFile}') }`,
        })
      } else {
        // If the Page doesn't exist then we are in a directory of Page
        // directories, so let's recurse into it and do the whole thing over
        // again.
        const newPrefix = [...prefix, entry.name]
        deps.push(
          ...processPagesDir(path.join(webPagesDir, entry.name), newPrefix)
        )
      }
    }
  })
  return deps
}
