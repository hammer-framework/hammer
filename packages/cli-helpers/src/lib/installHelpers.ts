import execa from 'execa'

import { getPaths } from './paths'

export const addWebPackages = (webPackages: string[]) => ({
  title: 'Adding required web packages...',
  task: async () => {
    await execa('yarn', ['add', ...webPackages], { cwd: getPaths().web.base })
  },
})

export const addApiPackages = (apiPackages: string[]) => ({
  title: 'Adding required api packages...',
  task: async () => {
    await execa('yarn', ['add', ...apiPackages], { cwd: getPaths().api.base })
  },
})

export const installPackages = {
  title: 'Installing packages...',
  task: async () => {
    await execa('yarn', ['install'], { cwd: getPaths().base })
  },
}
