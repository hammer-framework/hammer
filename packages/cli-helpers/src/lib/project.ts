import fs from 'fs'
import path from 'path'

import { colors } from './colors'
import { getPaths, resolveFile } from './paths'

export const getGraphqlPath = () => {
  return resolveFile(path.join(getPaths().api.functions, 'graphql'))
}

export const graphFunctionDoesExist = () => {
  const graphqlPath = getGraphqlPath()
  return graphqlPath && fs.existsSync(graphqlPath)
}

export const isTypeScriptProject = () => {
  const paths = getPaths()
  return (
    fs.existsSync(path.join(paths.web.base, 'tsconfig.json')) ||
    fs.existsSync(path.join(paths.api.base, 'tsconfig.json'))
  )
}

export const getInstalledRedwoodVersion = () => {
  try {
    const packageJson = require('../../package.json')
    return packageJson.version
  } catch (e) {
    console.error(colors.error('Could not find installed redwood version'))
    process.exit(1)
  }
}

export const addEnvVarTask = (name: string, value: string, comment: string) => {
  return {
    title: `Adding ${name} var to .env...`,
    task: () => {
      const envPath = path.join(getPaths().base, '.env')
      const content = [comment && `# ${comment}`, `${name}=${value}`, ''].flat()
      let envFile = ''

      if (fs.existsSync(envPath)) {
        envFile = fs.readFileSync(envPath).toString() + '\n'
      }

      fs.writeFileSync(envPath, envFile + content.join('\n'))
    },
  }
}
