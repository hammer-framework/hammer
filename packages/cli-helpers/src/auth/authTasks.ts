import fs from 'fs'
import path from 'path'

import execa from 'execa'
import { ListrRenderer, ListrTask, ListrTaskWrapper } from 'listr2'

import { transformTSToJS, writeFilesTask } from '../lib'
import { colors } from '../lib/colors'
import { getPaths, resolveFile } from '../lib/paths'
import {
  getGraphqlPath,
  graphFunctionDoesExist,
  isTypeScriptProject,
} from '../lib/project'

import { apiSideFiles, generateUniqueFileNames } from './authFiles'

const AUTH_PROVIDER_HOOK_IMPORT = `import { AuthProvider, useAuth } from './auth'`
const AUTH_HOOK_IMPORT = `import { useAuth } from './auth'`

export const getWebAppPath = () => getPaths().web.app

/**
 * This function looks for the createGraphQLHandler function call and adds
 * `authDecoder` to its arguments if it's not already there. Almost always it
 * will not be there, but if the user already has an auth provider set up and
 * wants to add another one it's probably there, and in that case we don't
 * want to add another one
 *
 * @param content - The contents of api/src/functions/graphql.ts
 * @returns content with `authDecoder` added unless it was already there
 */
function addAuthDecoderToCreateGraphQLHandler(content: string) {
  // Have to use a funky looking Regex here to prevent a "Polynomial regular
  // expression used on uncontrolled data" warning/error. A.k.a "Catastrophic
  // Backtracking". The usual fix is to use an atomic group, but the JS
  // regex engine doesn't support that, so we use a lookaround group to
  // emulate an atomic group.
  if (
    !new RegExp('(?=(^.*?createGraphQLHandler))\\1.*\\bauthDecoder', 's').test(
      content
    )
  ) {
    return content.replace(
      /^(?<indentation>\s*)(loggerConfig:)(.*)$/m,
      `$<indentation>authDecoder,\n$<indentation>$2$3`
    )
  }

  return content
}

/**
 * Replace the existing `import { authDecoder } from 'x'` with a new one
 *
 * @param content - The contents of api/src/functions/graphql.ts
 * @param decoderImport - Something like
 *   `import { authDecoder } from '@redwoodjs/auth-clerk-api'`
 * @returns content with the authDecoder import replaced with the new import
 */
function replaceAuthDecoderImport(content: string, decoderImport: string) {
  return content.replace(/^import { authDecoder .*} from .+/, decoderImport)
}

/**
 * Replace the existing `  authDecoder: myAuthDecoder,` with a standard
 * `  authDecoder,`
 *
 * @param content - The contents of api/src/functions/graphql.ts
 * @returns content with standard authDecoder arg to createGraphQLHandler
 */
function replaceAuthDecoderArg(content: string) {
  return content.replace(/^(\s+)authDecoder\b.+/m, '$1authDecoder,')
}

// Exporting this to make it easier to test
export const addApiConfig = (authDecoderImport?: string) => {
  const graphqlPath = getGraphqlPath()

  if (!graphqlPath) {
    throw new Error('Could not find your graphql file path')
  }

  const content = fs.readFileSync(graphqlPath).toString()
  let newContent = content

  // If there already is an import we replace it with a new one
  const replaceExistingImport = /^import { authDecoder .*} from /m.test(content)

  if (authDecoderImport) {
    if (replaceExistingImport) {
      newContent = replaceAuthDecoderImport(newContent, authDecoderImport)
      newContent = replaceAuthDecoderArg(newContent)
    } else {
      newContent = authDecoderImport + '\n' + newContent
      newContent = addAuthDecoderToCreateGraphQLHandler(newContent)
    }
  }

  const hasCurrentUserImport =
    /(^import {.*?getCurrentUser(?!getCurrentUser).*?} from 'src\/lib\/auth')/s.test(
      newContent
    )

  if (!hasCurrentUserImport) {
    // add import statement
    newContent = newContent.replace(
      /^(import { db } from 'src\/lib\/db')$/m,
      `import { getCurrentUser } from 'src/lib/auth'\n$1`
    )

    // add object to handler
    newContent = newContent.replace(
      /^(\s*)(loggerConfig:)(.*)$/m,
      `$1getCurrentUser,\n$1$2$3`
    )
  }

  if (newContent !== content) {
    fs.writeFileSync(graphqlPath, newContent)
  }
}

const apiSrcDoesExist = () => {
  return fs.existsSync(path.join(getPaths().api.src))
}

const addAuthImportToApp = (content: string) => {
  const contentLines = content.split('\n')
  // Find the last import line that's not a .css or .scss import
  // @ts-expect-error - We polyfill findLastIndex
  const importIndex = contentLines.findLastIndex((line: string) =>
    /^\s*import (?!.*(?:.css'|.scss'))/.test(line)
  )

  // After the import found above, insert a blank line followed by the
  // AuthProvider and useAuth import
  contentLines.splice(importIndex + 1, 0, '', AUTH_PROVIDER_HOOK_IMPORT)

  return contentLines.join('\n')
}

const addAuthImportToRoutes = (content: string) => {
  const contentLines = content.split('\n')
  // Find the last import line that's not a .css or .scss import
  // @ts-expect-error - We polyfill findLastIndex
  const importIndex = contentLines.findLastIndex((line: string) =>
    /^\s*import (?!.*(?:.css'|.scss'))/.test(line)
  )

  // After the import found above, insert a blank line followed by the
  // useAuth import
  contentLines.splice(importIndex + 1, 0, '', AUTH_HOOK_IMPORT)

  return contentLines.join('\n')
}

const hasAuthProvider = (content: string) => {
  return /\s*<AuthProvider>/.test(content)
}

/** returns the content of App.{js,tsx} with <AuthProvider> added */
const addAuthProviderToApp = (content: string) => {
  const match = content.match(
    /(\s+)(<RedwoodProvider.*?>)(.*)(<\/RedwoodProvider>)/s
  )

  if (!match) {
    throw new Error('Could not find <RedwoodProvider> in App.{js,tsx}')
  }

  const [
    _,
    newlineAndIndent,
    redwoodProviderOpen,
    redwoodProviderChildren,
    redwoodProviderClose,
  ] = match

  const redwoodProviderChildrenLines = redwoodProviderChildren
    .split('\n')
    .map((line, index) => {
      return `${index === 0 ? '' : '  '}` + line
    })

  const renderContent =
    newlineAndIndent +
    redwoodProviderOpen +
    newlineAndIndent +
    '  ' +
    `<AuthProvider>` +
    redwoodProviderChildrenLines.join('\n') +
    `</AuthProvider>` +
    newlineAndIndent +
    redwoodProviderClose

  return content.replace(
    /\s+<RedwoodProvider.*?>.*<\/RedwoodProvider>/s,
    renderContent
  )
}

const hasUseAuthHook = (componentName: string, content: string) => {
  return new RegExp(
    `<${componentName}.*useAuth={.*?}.*?>.*<\/${componentName}>`,
    's'
  ).test(content)
}

const addUseAuthHook = (componentName: string, content: string) => {
  return content.replace(
    `<${componentName}`,
    `<${componentName} useAuth={useAuth}`
  )
}

/**
 * Actually inserts the required config lines into App.{js,tsx}
 * Exported for testing
 */
export const addConfigToWebApp = () => {
  return {
    title: 'Updating web/src/App.{js,tsx}',
    task: () => {
      const webAppPath = getWebAppPath()

      if (!fs.existsSync(webAppPath)) {
        const ext = isTypeScriptProject() ? 'tsx' : 'js'
        throw new Error(`Could not find root App.${ext}`)
      }

      let content = fs.readFileSync(webAppPath).toString()

      if (!content.includes(AUTH_PROVIDER_HOOK_IMPORT)) {
        content = addAuthImportToApp(content)
      }

      if (!hasAuthProvider(content)) {
        content = addAuthProviderToApp(content)
      }

      if (/\s*<RedwoodApolloProvider/.test(content)) {
        if (!hasUseAuthHook('RedwoodApolloProvider', content)) {
          content = addUseAuthHook('RedwoodApolloProvider', content)
        }
      } else {
        console.warn(
          colors.warning(
            'Could not find <RedwoodApolloProvider>.\nIf you are using a custom ' +
              'GraphQL Client you will have to make sure it gets access to your ' +
              '`useAuth`, if it needs it.'
          )
        )
      }

      fs.writeFileSync(webAppPath, content)
    },
  }
}

export const createWebAuth = (
  basedir: string,
  provider: string,
  webAuthn: boolean
) => {
  return {
    title: 'Creating web/src/auth.{js,ts}',
    task: () => {
      const templatesBaseDir = path.join(basedir, 'templates', 'web')
      const templates = fs.readdirSync(templatesBaseDir)

      const templateFileName = templates.find((template) => {
        return template.startsWith('auth.' + (webAuthn ? 'webAuthn.ts' : 'ts'))
      })

      if (!templateFileName) {
        throw new Error('Could not find the auth.ts template')
      }

      const templateExtension = templateFileName.split('.').at(-2)

      // Find an unused filename
      // Start with web/src/auth.{ts,tsx}
      // Then web/src/providerAuth.{ts,tsx}
      // Then web/src/providerAuth2.{ts,tsx}
      // Then web/src/providerAuth3.{ts,tsx}
      // etc
      let authFileName = path.join(getPaths().web.src, 'auth')
      let i = 1
      while (resolveFile(authFileName)) {
        const count = i > 1 ? i : ''

        authFileName = path.join(getPaths().web.src, provider + 'Auth' + count)

        i++
      }

      authFileName = authFileName + '.' + templateExtension

      let template: string | undefined = fs.readFileSync(
        path.join(templatesBaseDir, templateFileName),
        'utf-8'
      )

      template = isTypeScriptProject()
        ? template
        : transformTSToJS(authFileName, template)

      fs.writeFileSync(authFileName, template)
    },
  }
}

export const addConfigToRoutes = () => {
  return {
    title: 'Updating Routes file...',
    task: () => {
      const webRoutesPath = getPaths().web.routes

      let content = fs.readFileSync(webRoutesPath).toString()

      if (!content.includes(AUTH_HOOK_IMPORT)) {
        content = addAuthImportToRoutes(content)
      }

      if (!hasUseAuthHook('Router', content)) {
        content = addUseAuthHook('Router', content)
      }

      fs.writeFileSync(webRoutesPath, content)
    },
  }
}

export const generateAuthApiFiles = <Renderer extends typeof ListrRenderer>(
  basedir: string,
  provider: string,
  force: boolean,
  webAuthn: boolean
): ListrTask<never, Renderer> => {
  return {
    title: 'Generating auth api side files...',
    task: (_ctx: never, task: ListrTaskWrapper<never, Renderer>) => {
      if (!apiSrcDoesExist()) {
        return task.skip?.('api/src not found, skipping')
      }

      // The keys in `filesRecord` are the full paths to where the file contents,
      // which is the values in `filesRecord`, will be written.
      let filesRecord = apiSideFiles({ basedir, webAuthn })

      if (!force) {
        const uniqueFilesRecord = generateUniqueFileNames(filesRecord, provider)

        if (
          Object.keys(filesRecord).join(',') !==
          Object.keys(uniqueFilesRecord).join(',')
        ) {
          console.warn(
            colors.warning(
              "To avoid overwriting existing files we've generated new file " +
                'names for the newly generated files. This probably means ' +
                `${provider} auth doesn't work out of the box. You'll most ` +
                'likely have to manually merge some of the generated files ' +
                'with your existing auth files'
            )
          )
        }

        filesRecord = uniqueFilesRecord
      }

      return writeFilesTask(filesRecord, { overwriteExisting: force })
    },
  }
}

export const addAuthConfigToGqlApi = <Renderer extends typeof ListrRenderer>(
  authDecoderImport?: string
) => ({
  title: 'Adding auth config to GraphQL API...',
  task: (_ctx: never, task: ListrTaskWrapper<never, Renderer>) => {
    if (graphFunctionDoesExist()) {
      addApiConfig(authDecoderImport)
    } else {
      task.skip?.('GraphQL function not found, skipping')
    }
  },
})

export const addWebPackages = (webPackages: string[]) => ({
  title: 'Adding required web packages...',
  task: async () => {
    await execa('yarn', ['workspace', 'web', 'add', ...webPackages])
  },
})

export const addApiPackages = (apiPackages: string[]) => ({
  title: 'Adding required api packages...',
  task: async () => {
    await execa('yarn', ['workspace', 'api', 'add', ...apiPackages])
  },
})

export const installPackages = {
  title: 'Installing packages...',
  task: async () => {
    await execa('yarn', ['install'])
  },
}
