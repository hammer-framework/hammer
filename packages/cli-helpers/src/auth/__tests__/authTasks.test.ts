// Have to use `var` here to avoid "Temporal Dead Zone" issues
// eslint-disable-next-line
var mockIsTypeScriptProject = true
global.__dirname = __dirname

jest.mock('../../lib/paths', () => {
  const path = require('path')

  return {
    ...jest.requireActual('../../lib/paths'),
    getPaths: () => {
      const base = '/mock/base/path'

      return {
        web: {
          src: path.join(base, 'web', 'src'),
        },
      }
    },
  }
})

jest.mock('../../lib/project', () => ({
  isTypeScriptProject: () => mockIsTypeScriptProject,
}))

jest.mock('../../lib', () => ({
  transformTSToJS: (_path: string, data: string) => data,
}))

// This will load packages/cli-helpers/__mocks__/fs.js
jest.mock('fs')

// @ts-expect-error - this is good enough for our mock
fs.readdirSync = () => {
  return ['auth.ts.template']
}

import fs from 'fs'
import path from 'path'

import { getPaths } from '../../lib/paths'
import { createWebAuth } from '../authTasks'

beforeEach(() => {
  mockIsTypeScriptProject = true

  // @ts-expect-error - This method is added by packages/cli-helpers/__mocks__/fs.js
  fs.__setMockFiles({
    '/mock/setup/path/templates/web/auth.ts.template': '// web auth template',
  })
})

it('writes an auth.ts file for TS projects', () => {
  const basedir = '/mock/setup/path'
  createWebAuth(basedir, 'auth0', false)

  expect(
    fs.readFileSync(path.join(getPaths().web.src, 'auth.ts'))
  ).toMatchSnapshot()
})

it('writes an auth.js file for JS projects', () => {
  mockIsTypeScriptProject = false
  const basedir = '/mock/setup/path'
  createWebAuth(basedir, 'auth0', false)

  expect(
    fs.readFileSync(path.join(getPaths().web.src, 'auth.js'))
  ).toMatchSnapshot()
})
