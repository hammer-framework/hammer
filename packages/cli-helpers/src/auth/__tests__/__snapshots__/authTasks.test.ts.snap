// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`writes an auth.js file for JS projects 1`] = `"// web auth template"`;

exports[`writes an auth.ts file for TS projects 1`] = `"// web auth template"`;
